Pagination function plural data

BBBmau opened the pull request:
Fixes https://github.com/hashicorp/terraform-provider-google/issues/19157

This simplifies the implementation of plural data-sources by creating a generic function that can be used regardless of the data-source.

The PR represents two uses of the function, with one simplifying implementation within an already existing List function and the second showcasing the function implementation within a Read function 

- [X] google_kms_crypto_keys - [pagination loop](https://github.com/GoogleCloudPlatform/magic-modules/blob/53302465ca4fafd09b0e44f2259b5290297258a5/mmv1/third_party/terraform/services/kms/data_source_google_kms_crypto_keys.go#L129-L171)
- [X] google_kms_key_rings  - [pagination loop](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/kms/data_source_google_kms_key_rings.go#L108-L149)
- [X] google_kms_crypto_key_versions - [pagination loop](https://github.com/GoogleCloudPlatform/magic-modules/blob/b1c656e86f2efce6b3043082e7e15147ce32723e/mmv1/third_party/terraform/services/kms/data_source_google_kms_crypto_key_versions.go#L170-L212)
- [X] google_storage_buckets - [pagination loop](https://github.com/hashicorp/terraform-provider-google-beta/blob/c45269d8a2805240fe84577251dc73a66c157924/google-beta/services/storage/data_source_google_storage_buckets.go#L70-L110)
- [x] google_storage_bucket_objects - [pagination loop](https://github.com/hashicorp/terraform-provider-google-beta/blob/c45269d8a2805240fe84577251dc73a66c157924/google-beta/services/storage/data_source_google_storage_bucket_objects.go#L73-L112)


<!--
Complete the self-review checklist to help speed up the review process: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

If your PR is still work in progress, please create it in draft mode.

Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to.
For example: Fixes https://github.com/hashicorp/terraform-provider-google/issues/ISSUE_ID
-->

**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:none

```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/45220a851f91547806d35906ee60eff59eba99ae..2cf7430a0150d6a9f99b62f67f1f75a549c7d819) ( 9 files changed, 321 insertions(+), 337 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/a916cc1d7c7d3d3bffb567bfc577fa9f7972b594..3da8eeacc7601ed4e5d61be1bff4c188957f5877) ( 9 files changed, 321 insertions(+), 337 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/98d3a4ff905bf0df7f604221c8f49e6ad63db2ae..21d5a95a45034acff2e1bee9ba343a870f76021e) ( 1 file changed, 137 insertions(+))





## Errors

`google` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

`google-beta` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/45220a851f91547806d35906ee60eff59eba99ae..bbe958ed95ba055912f9ceb60b442a70a0a1140a) ( 9 files changed, 321 insertions(+), 337 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/a916cc1d7c7d3d3bffb567bfc577fa9f7972b594..d161eaa7337d6ee496ccad0bb8dabd789d8503b4) ( 9 files changed, 321 insertions(+), 337 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/f7e2a4d7890bec3e4275e8464b56de5c2936962a..88084ec3d01b95a60caa747624d4a37bd089d9ee) ( 1 file changed, 137 insertions(+))





## Errors

`google` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

`google-beta` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

modular-magician commented:
#### Tests analytics
Total tests: 0
Passed tests: 0
Skipped tests: 0
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>
🔴 Errors occurred during REPLAYING mode. Please fix them to complete your PR.

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/819b539e-d5d2-4391-82be-725ca873d086/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/45220a851f91547806d35906ee60eff59eba99ae..429819cc50cf9f7eb5eb487ab0888b1c23e18296) ( 9 files changed, 330 insertions(+), 351 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/a916cc1d7c7d3d3bffb567bfc577fa9f7972b594..e4c367f956141c264d2e597c1679f029068b3b05) ( 9 files changed, 330 insertions(+), 351 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/1da65d1500e606c6b647ac8d1fc55d97ff7d55ed..757142494413b9b9c1fbb9577431d9d79c9e3eab) ( 1 file changed, 137 insertions(+))





## Errors

`google` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

`google-beta` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

modular-magician commented:
#### Tests analytics
Total tests: 0
Passed tests: 0
Skipped tests: 0
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>
🔴 Errors occurred during REPLAYING mode. Please fix them to complete your PR.

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/8ab894c2-e524-4fe2-a407-a693f4088a18/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/45220a851f91547806d35906ee60eff59eba99ae..40950416c32e89b7fe786f724692f10b4294fb0b) ( 9 files changed, 331 insertions(+), 338 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/a916cc1d7c7d3d3bffb567bfc577fa9f7972b594..cd3fd209d659780191fca764c8a16c5471c895d6) ( 9 files changed, 331 insertions(+), 338 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/aeec518904ede30749a07740c85e3edb38c5fc48..78bb42e6d378222ef634b03ac8f2b981ddc11829) ( 1 file changed, 137 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 5376
Passed tests: 4748
Skipped tests: 596
Affected tests: 32

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>
#### Action taken
<details>
<summary>Found 32 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccAccessContextManager__service_perimeter_dry_run_egress_policy</li>
<li>TestAccBigQueryExternalDataTable_CSV_WithSchemaAndConnectionIDAndHivePartitioning</li>
<li>TestAccContainerNodePool_EmptyGuestAccelerator</li>
<li>TestAccContainerNodePool_autoscaling</li>
<li>TestAccContainerNodePool_fastSocket</li>
<li>TestAccContainerNodePool_gcfsConfig</li>
<li>TestAccContainerNodePool_gvnic</li>
<li>TestAccContainerNodePool_privateRegistry</li>
<li>TestAccContainerNodePool_resize</li>
<li>TestAccContainerNodePool_resourceManagerTags</li>
<li>TestAccContainerNodePool_totalSize</li>
<li>TestAccContainerNodePool_version</li>
<li>TestAccContainerNodePool_withCgroupMode</li>
<li>TestAccContainerNodePool_withConfidentialNodes</li>
<li>TestAccContainerNodePool_withEnablePrivateNodesToggle</li>
<li>TestAccContainerNodePool_withHugepageConfig</li>
<li>TestAccContainerNodePool_withKubeletConfig</li>
<li>TestAccContainerNodePool_withLinuxNodeConfig</li>
<li>TestAccContainerNodePool_withLoggingVariantUpdates</li>
<li>TestAccContainerNodePool_withMachineAndDiskUpdate</li>
<li>TestAccContainerNodePool_withManagement</li>
<li>TestAccContainerNodePool_withMaxRunDuration</li>
<li>TestAccContainerNodePool_withNetworkConfig</li>
<li>TestAccContainerNodePool_withNodeConfig</li>
<li>TestAccContainerNodePool_withTaintsUpdate</li>
<li>TestAccContainerNodePool_withUpgradeSettings</li>
<li>TestAccContainerNodePool_withWindowsNodeConfig</li>
<li>TestAccContainerNodePool_withWorkloadIdentityConfig</li>
<li>TestAccDataSourceAlloydbLocations_basic</li>
<li>TestAccDialogflowEncryptionSpec_dialogflowEncryptionSpecBasicExample</li>
<li>TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample</li>
<li>TestAccLoggingFolderSettings_update</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccAccessContextManager__access_level` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__access_level.log)]
`TestAccAccessContextManager__access_level_condition` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__access_level_condition.log)]
`TestAccAccessContextManager__access_level_custom` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__access_level_custom.log)]
`TestAccAccessContextManager__access_level_full` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__access_level_full.log)]
`TestAccAccessContextManager__access_levels` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__access_levels.log)]
`TestAccAccessContextManager__access_policy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__access_policy.log)]
`TestAccAccessContextManager__access_policy_scoped` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__access_policy_scoped.log)]
`TestAccAccessContextManager__authorized_orgs_desc` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__authorized_orgs_desc.log)]
`TestAccAccessContextManager__gcp_user_access_binding` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__gcp_user_access_binding.log)]
`TestAccAccessContextManager__service_perimeter` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__service_perimeter.log)]
`TestAccAccessContextManager__service_perimeter_dry_run_egress_policy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__service_perimeter_dry_run_egress_policy.log)]
`TestAccAccessContextManager__service_perimeter_dry_run_ingress_policy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__service_perimeter_dry_run_ingress_policy.log)]
`TestAccAccessContextManager__service_perimeter_update` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__service_perimeter_update.log)]
`TestAccAccessContextManager__service_perimeters` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccAccessContextManager__service_perimeters.log)]
`TestAccContainerNodePool_EmptyGuestAccelerator` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_EmptyGuestAccelerator.log)]
`TestAccContainerNodePool_autoscaling` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_autoscaling.log)]
`TestAccContainerNodePool_fastSocket` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_fastSocket.log)]
`TestAccContainerNodePool_gcfsConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_gcfsConfig.log)]
`TestAccContainerNodePool_gvnic` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_gvnic.log)]
`TestAccContainerNodePool_privateRegistry` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_privateRegistry.log)]
`TestAccContainerNodePool_resize` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_resize.log)]
`TestAccContainerNodePool_totalSize` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_totalSize.log)]
`TestAccContainerNodePool_version` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_version.log)]
`TestAccContainerNodePool_withCgroupMode` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withCgroupMode.log)]
`TestAccContainerNodePool_withConfidentialNodes` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withConfidentialNodes.log)]
`TestAccContainerNodePool_withEnablePrivateNodesToggle` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withEnablePrivateNodesToggle.log)]
`TestAccContainerNodePool_withHugepageConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withHugepageConfig.log)]
`TestAccContainerNodePool_withKubeletConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withKubeletConfig.log)]
`TestAccContainerNodePool_withLinuxNodeConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withLinuxNodeConfig.log)]
`TestAccContainerNodePool_withLoggingVariantUpdates` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withLoggingVariantUpdates.log)]
`TestAccContainerNodePool_withMachineAndDiskUpdate` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withMachineAndDiskUpdate.log)]
`TestAccContainerNodePool_withManagement` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withManagement.log)]
`TestAccContainerNodePool_withNetworkConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withNetworkConfig.log)]
`TestAccContainerNodePool_withNodeConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withNodeConfig.log)]
`TestAccContainerNodePool_withTaintsUpdate` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withTaintsUpdate.log)]
`TestAccContainerNodePool_withUpgradeSettings` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withUpgradeSettings.log)]
`TestAccContainerNodePool_withWindowsNodeConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withWindowsNodeConfig.log)]
`TestAccContainerNodePool_withWorkloadIdentityConfig` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withWorkloadIdentityConfig.log)]
`TestAccDialogflowEncryptionSpec_dialogflowEncryptionSpecBasicExample` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccDialogflowEncryptionSpec_dialogflowEncryptionSpecBasicExample.log)]
`TestAccLoggingFolderSettings_update` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccLoggingFolderSettings_update.log)]


🔴 Tests failed when rerunning REPLAYING mode:
`TestAccAccessContextManager__service_perimeter_dry_run_egress_policy` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/replaying_build_after_recording/TestAccAccessContextManager_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/replaying_after_recording/TestAccAccessContextManager__service_perimeter_dry_run_egress_policy.log)]
`TestAccContainerNodePool_withNetworkConfig` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/replaying_build_after_recording/TestAccContainerNodePool_withNetworkConfig_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/replaying_after_recording/TestAccContainerNodePool_withNetworkConfig.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---


🔴 Tests failed during RECORDING mode:
`TestAccBigQueryExternalDataTable_CSV_WithSchemaAndConnectionIDAndHivePartitioning` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/recording_build/TestAccBigQueryExternalDataTable_CSV_WithSchemaAndConnectionIDAndHivePartitioning_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccBigQueryExternalDataTable_CSV_WithSchemaAndConnectionIDAndHivePartitioning.log)]
`TestAccContainerNodePool_resourceManagerTags` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/recording_build/TestAccContainerNodePool_resourceManagerTags_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_resourceManagerTags.log)]
`TestAccContainerNodePool_withMaxRunDuration` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/recording_build/TestAccContainerNodePool_withMaxRunDuration_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccContainerNodePool_withMaxRunDuration.log)]
`TestAccDataSourceAlloydbLocations_basic` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/recording_build/TestAccDataSourceAlloydbLocations_basic_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccDataSourceAlloydbLocations_basic.log)]
`TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/recording_build/TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording/TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/25058b52-7c01-4fe1-a21a-574d4727ea5d/recording) for each test

BBBmau commented:
for some reason the original PR was closed, had to make a new one. I addressed some of the comments you left from the previous PR let me know your thoughts. I'm specifically wondering about the handling of certain behaviors and leaving the user to handle them @rileykarson

rileykarson commented on mmv1/third_party/terraform/transport/transport.go:
```diff
@@ -189,3 +189,140 @@ func IsApiNotEnabledError(err error) bool {
 	return false
 }
+
+type GetPaginatedItemsSliceOptions struct {
```
nit: Can you document the fields on this, given it's a public function's interface?

rileykarson commented on mmv1/third_party/terraform/transport/transport.go:
```diff
@@ -189,3 +189,140 @@ func IsApiNotEnabledError(err error) bool {
+			RawURL:               url,
+			UserAgent:            paginationOptions.UserAgent,
+			Headers:              headers,
+			ErrorRetryPredicates: []RetryErrorPredicateFunc{Is429RetryableQuotaError},
```
Is it intentional that we're newly retrying this in some cases?

rileykarson commented on mmv1/third_party/terraform/transport/transport.go:
```diff
@@ -189,3 +189,140 @@ func IsApiNotEnabledError(err error) bool {
+	ResourceToList string
+}
+
+func GetPaginatedItemsSlice(paginationOptions GetPaginatedItemsSliceOptions) ([]interface{}, error) {
```
As far as I can tell this just returns a weaker type than GetPaginatedItemsMapOptions in all cases. Could we just strengthen the type in the callers? (We also appear to be weakening the type in AlloyDB locations right now)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/f25a1ae38a16a72702fca1c62198b4d44c15d0e8..832d4ecf17421f60b9631d17487e442d3a5f41d9) ( 10 files changed, 288 insertions(+), 367 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/d8f69f47c3ea3662c8d5b941c6a679b146aa4d6d..0ef791c3745bdc41f5f7b5d9a14cc084b61f2aa2) ( 10 files changed, 288 insertions(+), 367 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/8762dde452982c2c383b9b9f9fbdeaaac0e2ea0f..ce3c7f4e33d03279b551aaa0b1d36c75da1993bd) ( 1 file changed, 78 insertions(+))

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/f25a1ae38a16a72702fca1c62198b4d44c15d0e8..4f31e6f8edf846961487949bad3f4ba30c9ad493) ( 11 files changed, 298 insertions(+), 367 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/d8f69f47c3ea3662c8d5b941c6a679b146aa4d6d..8256827c17ffe1fbf25b00f39e796fcaeb62ec0f) ( 11 files changed, 298 insertions(+), 367 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/9a1b6f1c9a629a8c715af43eefd2d05f35e2dd4e..a1a3700b744fcc7907ed913a16b16b0999bed71f) ( 2 files changed, 88 insertions(+))





## Errors

`google` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

`google-beta` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/f25a1ae38a16a72702fca1c62198b4d44c15d0e8..fccadc4d2226c71b2176476a708f0f937e0f9978) ( 11 files changed, 314 insertions(+), 367 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/d8f69f47c3ea3662c8d5b941c6a679b146aa4d6d..a2e5b373743c654de9e9cd14f7545480c4e23c80) ( 11 files changed, 314 insertions(+), 367 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/e60661f42e3b65f450a5d6225b60f0b3c961f50b..5e8379942c0b464c9423099f239b807d12d7b8aa) ( 2 files changed, 104 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 5090
Passed tests: 4491
Skipped tests: 594
Affected tests: 5

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 5 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccAccessContextManager__service_perimeter_dry_run_egress_policy</li>
<li>TestAccContainerCluster_withNodeConfigKubeletConfigSettingsInNodePool</li>
<li>TestAccDataFusionInstance_dataFusionInstanceCmekExample</li>
<li>TestAccDataSourceGoogleStorageBucketObjects_basic</li>
<li>TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccAccessContextManager__access_level` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__access_level.log)]
`TestAccAccessContextManager__access_level_condition` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__access_level_condition.log)]
`TestAccAccessContextManager__access_level_custom` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__access_level_custom.log)]
`TestAccAccessContextManager__access_level_full` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__access_level_full.log)]
`TestAccAccessContextManager__access_levels` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__access_levels.log)]
`TestAccAccessContextManager__access_policy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__access_policy.log)]
`TestAccAccessContextManager__access_policy_scoped` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__access_policy_scoped.log)]
`TestAccAccessContextManager__authorized_orgs_desc` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__authorized_orgs_desc.log)]
`TestAccAccessContextManager__gcp_user_access_binding` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__gcp_user_access_binding.log)]
`TestAccAccessContextManager__service_perimeter` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__service_perimeter.log)]
`TestAccAccessContextManager__service_perimeter_dry_run_egress_policy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__service_perimeter_dry_run_egress_policy.log)]
`TestAccAccessContextManager__service_perimeter_dry_run_ingress_policy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__service_perimeter_dry_run_ingress_policy.log)]
`TestAccAccessContextManager__service_perimeter_update` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__service_perimeter_update.log)]
`TestAccAccessContextManager__service_perimeters` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccAccessContextManager__service_perimeters.log)]
`TestAccDataFusionInstance_dataFusionInstanceCmekExample` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccDataFusionInstance_dataFusionInstanceCmekExample.log)]
`TestAccDataSourceGoogleStorageBucketObjects_basic` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccDataSourceGoogleStorageBucketObjects_basic.log)]


🔴 Tests failed when rerunning REPLAYING mode:
`TestAccAccessContextManager__service_perimeter_dry_run_egress_policy` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/build-log/replaying_build_after_recording/TestAccAccessContextManager_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/replaying_after_recording/TestAccAccessContextManager__service_perimeter_dry_run_egress_policy.log)]
`TestAccDataFusionInstance_dataFusionInstanceCmekExample` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/build-log/replaying_build_after_recording/TestAccDataFusionInstance_dataFusionInstanceCmekExample_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/replaying_after_recording/TestAccDataFusionInstance_dataFusionInstanceCmekExample.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---


🔴 Tests failed during RECORDING mode:
`TestAccContainerCluster_withNodeConfigKubeletConfigSettingsInNodePool` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/build-log/recording_build/TestAccContainerCluster_withNodeConfigKubeletConfigSettingsInNodePool_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccContainerCluster_withNodeConfigKubeletConfigSettingsInNodePool.log)]
`TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/build-log/recording_build/TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording/TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-14687/artifacts/4efbe2bc-52ff-494e-9056-8b62fdc375bf/recording) for each test

modular-magician commented:
#### Tests analytics
Total tests: 5030
Passed tests: 4432
Skipped tests: 594
Affected tests: 4

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 4 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeRegionBackendService_regionBackendServiceHaPolicyManualLeader_update</li>
<li>TestAccDataFusionInstance_dataFusionInstanceCmekExample</li>
<li>TestAccDialogflowEncryptionSpec_dialogflowEncryptionSpecBasicExample</li>
<li>TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
#### Tests analytics
Total tests: 5434
Passed tests: 4804
Skipped tests: 594
Affected tests: 36

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 36 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccAccessContextManager__service_perimeter_dry_run_egress_policy</li>
<li>TestAccContainerCluster_withNodeConfigKubeletConfigSettingsInNodePool</li>
<li>TestAccContainerNodePool_EmptyGuestAccelerator</li>
<li>TestAccContainerNodePool_autoscaling</li>
<li>TestAccContainerNodePool_fastSocket</li>
<li>TestAccContainerNodePool_gcfsConfig</li>
<li>TestAccContainerNodePool_gvnic</li>
<li>TestAccContainerNodePool_privateRegistry</li>
<li>TestAccContainerNodePool_regionalAutoscaling</li>
<li>TestAccContainerNodePool_resize</li>
<li>TestAccContainerNodePool_resourceManagerTags</li>
<li>TestAccContainerNodePool_totalSize</li>
<li>TestAccContainerNodePool_version</li>
<li>TestAccContainerNodePool_withCgroupMode</li>
<li>TestAccContainerNodePool_withConfidentialNodes</li>
<li>TestAccContainerNodePool_withEnablePrivateNodesToggle</li>
<li>TestAccContainerNodePool_withHugepageConfig</li>
<li>TestAccContainerNodePool_withKubeletConfig</li>
<li>TestAccContainerNodePool_withLegacyNodeConfigAndBootDiskUpdate</li>
<li>TestAccContainerNodePool_withLinuxNodeConfig</li>
<li>TestAccContainerNodePool_withLoggingVariantUpdates</li>
<li>TestAccContainerNodePool_withMachineAndDiskUpdate</li>
<li>TestAccContainerNodePool_withManagement</li>
<li>TestAccContainerNodePool_withMaxRunDuration</li>
<li>TestAccContainerNodePool_withNetworkConfig</li>
<li>TestAccContainerNodePool_withNodeConfig</li>
<li>TestAccContainerNodePool_withNodeConfigWithBootDiskConfig</li>
<li>TestAccContainerNodePool_withNodeConfigWithBootDiskConfigChangeType</li>
<li>TestAccContainerNodePool_withNodeConfigWithBootDiskConfigChangeTypeLegacy</li>
<li>TestAccContainerNodePool_withTaintsUpdate</li>
<li>TestAccContainerNodePool_withUpgradeSettings</li>
<li>TestAccContainerNodePool_withWindowsNodeConfig</li>
<li>TestAccContainerNodePool_withWorkloadIdentityConfig</li>
<li>TestAccDataFusionInstance_dataFusionInstanceCmekExample</li>
<li>TestAccHealthcarePipelineJob_healthcarePipelineJobBackfillExample</li>
<li>TestAccLoggingFolderSettings_update</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)
